packages/ifpack2/src/Ifpack2_Relaxation_def.hpp

```diff
@@ -634,19 +635,15 @@ void Relaxation<MatrixType>::initialize ()
     }
 
     if (PrecType_ == Ifpack2::Details::MTGS || PrecType_ == Ifpack2::Details::MTSGS) {
-      const crs_matrix_type* crsMat = dynamic_cast<const crs_matrix_type*> (&(*A_));
+      const crs_matrix_type* crsMat = dynamic_cast<const crs_matrix_type*> (A_.get());
       TEUCHOS_TEST_FOR_EXCEPTION
         (crsMat == NULL, std::logic_error, "Ifpack2::Relaxation::initialize: "
          "Multithreaded Gauss-Seidel methods currently only work when the input "
          "matrix is a Tpetra::CrsMatrix.");
 
       if(this->ifpack2_dump_matrix_){
-        int random_integer = rand();
-        std::stringstream ss;
-        ss << random_integer;
-        std::string str = ss.str();
+        std::string file_name = "Ifpack2_MT_GS." + std::to_string(::getpid()) + ".mtx";
```

Comment:
This is both wrong, and ineffective.

  1. Different MPI processes may (or may not!) have different OS process IDs.
  2. Tpetra::MatrixMarket::Writer is a collective, and it writes a single file.  All processes in the matrix's communicator need to agree on the filename.

Do this instead:
```c++
if (this->ifpack2_dump_matrix_) {
  static int sequence_number = 0;
  const std::string file_name = "Ifpack2_MT_GS_" + std::to_string (sequence_number++) + ".mtx";
  // ... etc. ...
}
```
Also, we really don't want to use POSIX stuff not in the C++ standard unless there's no other way.